### PR TITLE
lxd-generate: Refactor `stmt` generation with more manual intervention

### PR DESCRIPTION
**Pull request by masnax**
Depends on #10870 

This reworks a lot of the logic for generating SQL statements using `lxd-generate`. 

* Firstly, a new `sql=<table>.<name>` tag has been added, which allows per-field assignment of table and name. 
* The `via` tag has been removed. Instead, indirect joins can now be specified through the `joinon` tag. This works as follows, using `Snapshot` as an example:

Snapshots have a Project field, which is populated by two joins: A join on `instances`, by comparing the `instance_id` foreign key, and a join on `projects`, using the `project_id` foreign key from the `instances` table.

```sql
SELECT ...
FROM instances_snapshots
JOIN instances ON instances_snapshots.instance_id = instances.id
JOIN projects ON instances.project_id = projects.id
```

Using the `joinon` tag, we can set the `Projects` field to formulate a join on `instances.project_id`:
```go
type InstanceSnapshot struct {
  Project      string `db:"join=projects.name&joinon=instances.project_id"`
  Instance    string `db:"join=instances.name"`
}
```

`lxd-generate` will notice that the `<table>` from the `join` tag and the `joinon` tag do not match, and will assume that this is an indirect join, and thus has no column on the `instances_snapshots` table, and will create the clauses appropriately.

* Lastly, the `--database` flag for specifying the package the database helpers (`RegisterStmt`, `StmtString`, etc.) are in now expects a full import path `"github.com/path/to/package"`, rather than just the package name, and hoping `goimports` will find the right package. 
---

Additionally, to facilitate the `sql` tag, and also make the `stmt.go` file a bit cleaner, there has been a substantial refactor to the `stmt.go` file. Rather than having a bunch of duplicated logic all around the `stmt.go` file, I've now added methods on `Mapping` and `Field` that are called within `stmt.go` to translate a field / struct to its database equivalent. 

These methods are:
* `mapping.TableName` - This determines the default table name. The table name can be overwritten with the `table=<name>` config key, else it will be generated from the struct name. Any individual field will be able to bypass this default table name with the `sql` tag described above.

* `field.SelectColumn` - This will create a column suitable for a `SELECT` statement from the information we have about the field. It will take into consideration the `join` tag and the `sql` tag to produce a result of the form `<table>.<column>`, or `<joinTable>.<joinColumn> AS <sqlColumn>` if a join is specified.

* `field.OrderBy` - This will create an `ORDER BY` clause, taking into consideration the `primary_key`, `orderby`, `sql`, and `join` tags. If the `orderby` tag is present, then only fields with the `orderby` tag will be used. Otherwise, all `primary_key` fields will be included in the clause. Field-to-column translation will be determined by `sql` and `join` tags.

* `field.InsertColumn` - This will create a column and its associated value parameter, suitable for an `INSERT`, `UPDATE`, or `DELETE` statement. If a `join` tag is present, then the column name will be of the form `<joinTable_id>` unless overridden by the `sql` tag. The value parameter will then be replaced with the `SELECT` statement produced by the `id` statement directive for the joining table. This will be of the form `var <joining entity>ID = <database>.RegisterStmt("SQL statement goes here")`. This makes `lxd-generate` a bit more self referential, but I think this is better than adding further complexity in trying to extrapolate the structure of the `SELECT` statement from the information in the `join` tag, like we were doing before. To find this field, the package that the struct entity is defined in will be checked first. If the variable does not exist there, and the `--database` flag is given, then that package is checked as well before failing.

* `field.JoinClause` - This will create a `JOIN` or `LEFT JOIN` clause of the form `JOIN <table> ON <table>.<joinTable>_id = <joinTable>.id`. `<table>` can be overridden by the `sql` tag, and `<table>.<joinTable>_id` can be overridden entirely by the `joinon` tag.

### CODE REVIEW

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by masnax** [comment]
@tomponline Done.

**Review by tomponline — CHANGES_REQUESTED** [review]
The `lxd/db/generate/db/stmt: Rework statement generation ` commit has a "Squash rework" comment and double sign-off in the commit message.

**Comment by masnax** [comment]
Ah, darnit! Fixed now. Sorry about that...

File: lxd/db/cluster/nodes.mapper.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+// GetNodeID return the ID of the node with the given key.
+// generator: node ID
+func GetNodeID(ctx context.Context, tx *sql.Tx, name string) (int64, error) {
+	stmt, err := Stmt(tx, nodeID)
+	if err != nil {
+		return -1, fmt.Errorf("Failed to get \"nodeID\" prepared statement: %w", err)
+	}
+
+	rows, err := stmt.Query(name)
```
Comment (tomponline) [inline lxd/db/cluster/nodes.mapper.go:31]: Why dont we use https://pkg.go.dev/database/sql#Tx.QueryRowContext here? For two reason, firstly we would be passing the context into the query function itself, secondly this function will return at most 1 rows or error if now rows found. The `ErrNoRows` error can be converted into the `api.StatusErrorf(http.StatusNotFound, "Node not found")` error using `errors.Is(err, sql.ErrNoRows)`.

This got me thinking too, do we use https://pkg.go.dev/database/sql#Tx.QueryContext in the DB generator when we do expect multiple rows?

**Review by tomponline — APPROVED** [review]
Looks good to me, we can fix the QueryRowContext and QueryContext calls in a separate PR.
I like the approach of moving the SQL building logic into their own functions to avoid duplication of logic.

**Comment by tomponline** [comment]
> Ah, darnit! Fixed now. Sorry about that...

I dont think you pushed to this PR though.

**Pull request merged**